src/Http/Environment.php

```diff
@@ -0,0 +1,904 @@
+                continue;
+            }
+
+            $uri = new Uri($url, ['slash' => false]);
+
+            if ($uri->toString() === $this->url) {
+                return;
+            }
+        }
+
+        throw new InvalidArgumentException('The environment is not allowed');
+    }
+
+    /**
+     * The URL option receives a set of Server constant flags
+     *
+     * Server::HOST_FROM_SERVER
+     * Server::HOST_FROM_SERVER | Server::HOST_ALLOW_EMPTY
+     * Server::HOST_FROM_HOST
+     * Server::HOST_FROM_HOST | Server::HOST_ALLOW_EMPTY
+     *
+     * @param int $flags
+     * @return string|null
+     */
+    protected function detectAllowedFromFlag(int $flags): ?string
+    {
+        // allow host detection from host headers
+        if ($flags & Server::HOST_FROM_HEADER) {
+            return '*';
+        }
+
+        // detect host only from server name
+        return null;
+    }
+
+    /**
+     * @param bool $insecure
+     * @return void
+     */
+    protected function detectAuto(bool $insecure = false): void
+    {
+        // proxy server setup
+        if ($this->isBehindProxy === true && $insecure === true) {
+            $this->host  = $this->detectForwardedHost();
+            $this->https = $this->detectForwardedHttps();
+            $this->port  = $this->detectForwardedPort();
+
+        // local server setup
+        } else {
+            $this->host  = $this->detectHost($insecure);
+            $this->https = $this->detectHttps();
+            $this->port  = $this->detectPort();
+        }
+    }
+
+    /**
+     * @param bool|null $override
+     * @return bool
+     */
+    protected function detectCli(?bool $override = null): bool
+    {
+        if ($override === false) {
+            return false;
+        }
+
+        if ($override === true) {
+            return true;
+        }
+
+        if (defined('STDIN') === true) {
+            return true;
+        }
+
+        $term = getenv('TERM');
+
+        if (substr(PHP_SAPI, 0, 3) === 'cgi' && $term && $term !== 'unknown') {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @return string|null
+     */
+    protected function detectForwardedHost(): ?string
+    {
+        $host = $this->get('HTTP_X_FORWARDED_HOST');
+
+        if (empty($host) === false) {
+            $parts = $this->detectPortInHost($host);
+
+            $this->portInHost = $parts['port'];
+
+            return $parts['host'];
+        }
+
+        return $this->detectHost(true);
+    }
+
+    /**
+     * @return bool
+     */
+    protected function detectForwardedHttps(): bool
+    {
+        if ($this->detectHttpsOn($this->get('HTTP_X_FORWARDED_SSL')) === true) {
+            return true;
+        }
+
+        if ($this->detectHttpsProtocol($this->get('HTTP_X_FORWARDED_PROTO')) === true) {
+            return true;
+        }
+
+        // fall back to local setup
+        return $this->detectHttps();
```

Comment:
TBH I'm not experienced enough with proxy setups. If the local setup really does not provide any useful reference in such a case, we can really skip it. Wouldn't it be fine to just assume it's not HTTPS in that case instead of throwing an exception?